modules/api/php/candidatesprovisioner.class.inc

```diff
@@ -0,0 +1,91 @@
+<?php declare(strict_types=1);
+/**
+ * This file implements a data provisioner to get the candidate information
+ * of all active candidates accessible to a given user.
+ *
+ * PHP Version 7
+ *
+ * @category API
+ * @package  Loris
+ * @author   Xavier Lecours Boucher <xavier.lecours@mcin.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+
+namespace LORIS\api;
+
+/**
+ * This class implements a data provisioner to get the candidate information
+ * of all active candidates accessible to a given user.
+ *
+ * PHP Version 7
+ *
+ * @category API
+ * @package  Loris
+ * @author   Xavier Lecours Boucher <xavier.lecours@mcin.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+class CandidatesProvisioner extends \LORIS\Data\Provisioners\DBRowProvisioner
+{
+    /**
+     * Create a RowProvisioner, which gets rows for candidates/
+     * endpoint.
+     */
+    function __construct()
+    {
+        parent::__construct(
+            '
+             SELECT 
+               c.CandID as CandID,
+               p.Name as ProjectName,
+               c.PSCID as PSCID,
+               c.RegistrationCenterID as CenterID,
+               s.Name as SiteName,
+               c.EDC as EDC,
+               c.DoB as DoB,
+               c.Sex as Sex
+             FROM
+               candidate c
+             JOIN psc s
+               ON (s.CenterID = c.RegistrationCenterID)
+             LEFT JOIN Project p
+               ON (c.ProjectID = p.ProjectID)
+             WHERE c.Active="Y"
+            ',
+            array()
+        );
+    }
+
+    /**
+     * Creates a new providionner with user specific filter.
+     *
+     * @param \User $user The requesting user
+     *
+     * @return CandidatesProvisioner
+     */
+    public function forUser(\User $user): CandidatesProvisioner
```

Comment:
I'm not sure about this factory being the right way to do this, vs just unconditionally adding a filter and doing the permission check inside of the (new) filter, since the filter interface is already passed the User object and it seems to be a more appropriate place to handle the concern.

Reply:
> doing the permission check inside of the (new) filter

 Would that mean to check permission for each row?

Originally, I wanted to pass the User in the constructor but I was shy to change `DBRowProvisioner` constructor signature. Having a provisioner know which filter to apply seemed like a good segregation of concerns as oppose to asking the user to add arbitrary filters. 

_On a side note_, I think that the filter function should be chnaged so that it adds expressions to the `WHERE` clause of the query. Databases are usually good at filtering rows, and it would reduce memory usage by reducing the result size.

Reply:
> Would that mean to check permission for each row?

Yes, but the permissions are currently loaded by the User constructor. Checking permissions is an O(1) array lookup, not a database call.

> On a side note, I think that the filter function should be chnaged so that it adds expressions to the WHERE clause of the query. Databases are usually good at filtering rows, and it would reduce memory usage by reducing the result size.

It's unlikely to have a significant affect on memory usage since DBRowProvisioner uses an iterator which lazily converts to the respective DataInstance class on iteration. The memory overhead should be approximately the size of 1 row, plus a little overhead for garbage collection regardless of result set size. (It's the main reason it's more efficient than $DB->pselect, which returns an array that must be fully loaded into memory.) If memory has went up to higher than that, it's a regression that should be fixed.